pkg/systemd/init.js

```diff
@@ -0,0 +1,745 @@
+     *   so calling GetAll as a reaction to UnitNew might lead to
+     *   infinite loops.
+     *
+     * - To avoid this cycle, we only call GetAll when there is some
+     *   job activity for a unit, or when the whole daemon is
+     *   reloaded.  The idea is that without jobs or a full reload,
+     *   the state of a unit will not change in an interesting way.
+     *
+     * - We hope that the cache machinery in cockpit-bridge does not
+     *   trigger such a cycle when watching a unit.
+     *
+     * - JobNew and JobRemoved signals don't include the object path
+     *   of the affected units, but we can get those by listening to
+     *   UnitNew.
+     *
+     * - There might be UnitNew signals for units that are never
+     *   returned by ListUnits or ListUnitFiles.  These are units that
+     *   are mentioned in Requires, After, etc or that people try to
+     *   load via LoadUnit but that don't actually exist.
+     *
+     * - ListUnitFiles will return unit files that are aliases for
+     *   other unit files, but ListUnits will not return aliases.
+     *
+     * - The "Names" property of a unit only includes those aliases
+     *   that are currently loaded, not all.  To get all possible
+     *   aliases, one needs to call ListUnitFiles and match units via
+     *   their object path.
+     *
+     * - The unit file state of a alias as returned by ListUnitFiles
+     *   is always the same as the unit file state of the primary unit
+     *   file.
+     *
+     * - However, the unit file state as returned by ListUnitFiles is
+     *   not necessarily the same as the UnitFileState property of a
+     *   loaded unit.  ListUnitFiles reflects the state of the files
+     *   on disk, while a loaded unit is only updated to that state
+     *   via an explicit Reload.
+     *
+     * - Thus, we are careful to only use the UnitFileState as
+     *   returned by ListUnitFiles or GetUnitFileState.  The
+     *   alternative would be to only use the UnitFileState property,
+     *   but we need one method call per unit to get them all for the
+     *   overview, which seems excessive.
+     *
+     * - Methods like EnableUnitFiles only change the state of files
+     *   on disk.  A Reload is necessary to update the state
+     *   of loaded units.
+     *
+     * - A Reload will emit UnitRemoved/UnitNew signals for all units,
+     *   and no PropertiesChanges signal for the properties that have
+     *   changed because of the reload, such as UnitFileState.
+     *
+     */
+
+    function startsWith(string, prefix) {
+        return string.indexOf(prefix) === 0;
+    }
+
+    /* See systemd-escape(1), used for instantiating templates.
+     */
+
+    function systemd_escape(str) {
+
+        function name_esc(str) {
+            var validchars = /[0-9a-zA-Z:-_.\\]/;
+            var res = "";
+            var i;
+
+            for (i = 0; i < str.length; i++) {
```

Comment:
If this code is used a *lot* we may want to assign str.length to a variable. If this code is used rarely, ignore my silliness.